security/pkg/nodeagent/cache/secretcache.go

```diff
@@ -335,6 +323,52 @@ func (sc *SecretCache) GenerateSecret(ctx context.Context, connectionID, resourc
 	return ns, nil
 }
 
+func (sc *SecretCache) addFileWatcher(file string, token string, connKey ConnKey) {
+	// TODO(ramaraochavali): add integration test for file watcher functionality.
+	// Check if this file is being already watched, if so ignore it. FileWatcher has the functionality of
+	// checking for duplicates. This check is needed here to avoid processing duplicate events for the same file.
+	sc.certMutex.Lock()
+	_, exists := sc.fileCerts[file]
+	if !exists {
+		sc.fileCerts[file] = struct{}{}
+	}
+	sc.certMutex.Unlock()
+	// File is not being watched, start watching now and trigger key push.
+	if !exists {
+		cacheLog.Debugf("adding watcher for file %s", file)
+		if err := sc.certWatcher.Add(file); err != nil {
+			cacheLog.Errorf("error adding watcher for file %s, Skipping watches ", file)
+			return
+		}
+		go func() {
+			var timerC <-chan time.Time
+			for {
+				select {
+				case <-timerC:
+					timerC = nil
+					// TODO(ramaraochavali): Remove the watchers for unused keys and certs.
+					if _, ok := sc.secrets.Load(connKey); ok {
+						// Regenerate the Secret and trigger the callback that pushes the secrets to proxy.
+						if _, secret, err := sc.generateFileSecret(connKey, token); err != nil {
+							cacheLog.Errorf("error in generating secret after file change %s, %v", file, err)
+						} else {
+							cacheLog.Debugf("file changed %s, triggering push to proxy", file)
+							sc.callbackWithTimeout(connKey, secret)
+						}
+					}
+				case e := <-sc.certWatcher.Events(file):
+					if len(e.Op.String()) > 0 { // To avoid spurious events, mainly coming from tests.
+						// Use a timer to debounce watch updates
```

Comment:
Sorry, can you help me to understand how this debounce solves the spurious events issue? If every event is delated 100ms in processing, we still have a burst of events, just 100ms later, no?

Reply:
This debounce does not solve the spurious events case. Spurious events are mainly coming from tests injection. What this debounce solves is, let us say if key is modified, we will get an event, then later cert is modified we will get another event - This debounce helps to combine both events in to one and generate secret item once and push it on the SDS connection once.

Reply:
Got it. You are nullifying timerC to do the debounce. Thanks!